File: app/src/main/java/com/popalay/yocard/ui/debts/AddDebtActivity.java
package com.popalay.yocard.ui.debts;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.transition.Transition;
import android.view.View;

import com.popalay.yocard.R;
import com.popalay.yocard.databinding.ActivityAddDebtBinding;
import com.popalay.yocard.ui.base.BaseActivity;
import com.popalay.yocard.ui.transitions.FabTransform;
import com.popalay.yocard.ui.transitions.MorphTransform;
import com.popalay.yocard.utils.TransitionUtils;

public class AddDebtActivity extends BaseActivity {

    private static final String TAG = "AddDebtActivity";

    private ActivityAddDebtBinding b;

    public static Intent getIntent(Context context) {
        return new Intent(context, AddDebtActivity.class);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        b = DataBindingUtil.setContentView(this, R.layout.activity_add_debt);

        if (!FabTransform.setup(this, b.container)) {
            MorphTransform.setup(this, b.container,
                    ContextCompat.getColor(this, R.color.dialog_background),
                    getResources().getDimensionPixelSize(R.dimen.dialog_corners));
        }
        if (getWindow().getSharedElementEnterTransition() != null) {
            getWindow().getSharedElementEnterTransition().addListener(new TransitionUtils.TransitionListenerAdapter() {
                @Override
                public void onTransitionEnd(Transition transition) {
                    getWindow().getSharedElementEnterTransition().removeListener(this);
                    initUI();
                }
            });
        } else {
            initUI();
        }
    }

    @Override
    public void onBackPressed() {
        dismiss(null);
    }

    public void dismiss(View view) {
        setResult(Activity.RESULT_CANCELED);
        finishAfterTransition();
    }

    private void initUI() {
        b.buttonSave.setOnClickListener(v -> close());
    }
}

File: presentation/src/main/kotlin/com/popalay/cardme/screens/adddebt/AddDebtActivity.kt
package com.popalay.cardme.screens.adddebt

import android.arch.lifecycle.ViewModelProvider
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.widget.AutoCompleteTextView
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.ScrollView
import com.jakewharton.rxbinding2.view.RxView
import com.jakewharton.rxbinding2.widget.RxTextView
import com.popalay.cardme.R
import com.popalay.cardme.base.BaseActivity
import com.popalay.cardme.base.mvi.BindableMviView
import com.popalay.cardme.base.navigation.CustomNavigator
import com.popalay.cardme.base.navigation.CustomRouter
import com.popalay.cardme.screens.setDropDownItems
import com.popalay.cardme.utils.extensions.applyThrottling
import com.popalay.cardme.utils.extensions.bindView
import com.popalay.cardme.utils.extensions.getViewModel
import com.popalay.cardme.utils.extensions.hideAnimated
import com.popalay.cardme.utils.extensions.onEnd
import com.popalay.cardme.utils.extensions.setTextIfNeeded
import com.popalay.cardme.utils.extensions.showAnimated
import com.popalay.cardme.utils.transitions.FabTransform
import io.reactivex.Observable
import javax.inject.Inject

class AddDebtActivity : BaseActivity(), BindableMviView<AddDebtViewState, AddDebtIntent> {

    companion object {

        fun getIntent(context: Context) = Intent(context, AddDebtActivity::class.java)
    }

    private val inputTo: AutoCompleteTextView by bindView(R.id.input_to)
    private val inputMessage: EditText by bindView(R.id.input_message)
    private val container: LinearLayout by bindView(R.id.container)
    private val buttonSave: FloatingActionButton by bindView(R.id.button_save)
    private val layoutRoot: ScrollView by bindView(R.id.layout_root)

    @Inject lateinit var factory: ViewModelProvider.Factory
    @Inject override lateinit var navigator: CustomNavigator
    @Inject lateinit var router: CustomRouter

    private lateinit var lastState: AddDebtViewState

    private val initialIntent = Observable.just(AddDebtIntent.Initial.GetHolderNames)

    private val nameChangeIntents = RxTextView.textChanges(inputTo)
        .skipInitialValue()
        .map { it.toString() }
        .map { lastState.debt.copy(holderName = it) }
        .map(AddDebtIntent::DebtHolderNameChanged)

    private val informationChangeIntents = RxTextView.textChanges(inputMessage)
        .skipInitialValue()
        .map { it.toString() }
        .map { lastState.debt.copy(message = it) }
        .map(AddDebtIntent::DebtInformationChanged)

    private val saveIntents = RxView.clicks(buttonSave)
        .applyThrottling()
        .map { lastState.debt }
        .map(AddDebtIntent::Accept)

    override val intents: Observable<AddDebtIntent> = Observable.defer {
        Observable.merge(
            initialIntent,
            informationChangeIntents,
            nameChangeIntents,
            saveIntents
        )
    }

    override fun accept(state: AddDebtViewState) {
        lastState = state
        with(state) {
            with(debt) {
                inputTo.setTextIfNeeded(holderName)
                inputMessage.setTextIfNeeded(message)
            }
            inputTo.setDropDownItems(holderNames)
            buttonSave.isEnabled = canSave
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_debt)
        initUi()

        bind(getViewModel<AddDebtViewModel>(factory))
    }

    override fun onBackPressed() = exitWithAnimation()

    fun exitWithAnimation() {
        buttonSave.hideAnimated { supportFinishAfterTransition() }
    }

    private fun initUi() {
        FabTransform.setup(this, container)
        window.sharedElementEnterTransition.onEnd { buttonSave.showAnimated() }
        layoutRoot.setOnClickListener { router.exit() }
    }
}
